### PR TITLE
EPD-11732 Issue 1772/open calendar on available dates

### PR DESCRIPTION
**Pull request by yevhenkravets**
## Description

If the date input is being used with a picker and min and max properties have been set to the input and the user typed a date to the input that is out of the min-max range then when the user clicks on the calendar icon then open a calendar widget on the month that has available dates and is closer to the min-max date.

When the selected date is before the min-max range - then open a calendar on the min date.
When the selected date is after the min-max range - then open a calendar on the max date.

References #1772 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## How Has This Been Tested?

This will be tested by e-builder QA. Also checked all scenarios through the integration tests.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: stencil-workspace/src/components/modus-date-picker/modus-date-picker.tsx [hunk 1]
```diff
@@ -157,7 +157,22 @@ export class ModusDatePicker {
     }
 
     const date = pickedDate || new Date();
-    this._calendar.gotoDate(date.getFullYear(), date.getMonth());
+
+    // if picked date is out od min-max range - open calendar on closer available month to pick
```
Comment (cjwinsor) [inline stencil-workspace/src/components/modus-date-picker/modus-date-picker.tsx:161]: typo with `is out od`

**Comment by cjwinsor** [comment]
@yevhenkravets I think what was missed in past PRs is that the min and max aren't watched. So as I use the canvas to test this PR, and I changed the min and max dates, the state only knows the original min and max.  If you change the min date and then change the date to be before that, and open the calendar then it will load the calendar based on the original min date.

Can we update min and max so can be updated after initial constructed.

File: stencil-workspace/src/components/modus-date-picker/modus-date-picker.tsx [hunk 2]
```diff
@@ -157,7 +157,22 @@ export class ModusDatePicker {
...
+    const min = this._currentInput?.getMinDateAllowed();
+
+    if (this.compare(date, min) < 0) {
+      this._calendar.gotoDate(min.getFullYear(), min.getMonth());
+      return;
+    }
+
+    const max = this._currentInput?.getMaxDateAllowed();
+    this._calendar.gotoDate(max.getFullYear(), max.getMonth());
   }
```
Comment (cjwinsor) [inline stencil-workspace/src/components/modus-date-picker/modus-date-picker.tsx:176]: We should consider breaking this up a little so the functions are a little more focused and some of the slight duplication is reduced

```js
  gotoDateBeingPicked(pickedDate: Date): void {
    if (!this._showCalendar) return;

    const date = pickedDate ?? new Date();
    
    if (this.isWithinCurrentMinMax(date)) {
      this.goToCalendarDate(date);
      return;
    }

    this.goToNearestBoundaryDate(date);
  }

  private goToCalendarDate(date: Date): void {
    this._calendar.gotoDate(date.getFullYear(), date.getMonth());
  }

  private goToNearestBoundaryDate(date: Date): void {
    const minDate = this._currentInput?.getMinDateAllowed();
    const maxDate = this._currentInput?.getMaxDateAllowed();

    const targetDate = this.compare(date, minDate) < 0 ? minDate : maxDate;
    this.goToCalendarDate(targetDate);
  }
```

File: stencil-workspace/src/components/modus-date-picker/modus-date-picker.tsx (hunk 1 above)
Comment (yevhenkravets) [reply stencil-workspace/src/components/modus-date-picker/modus-date-picker.tsx:161]: Thank you! Will update it

File: stencil-workspace/src/components/modus-date-picker/modus-date-picker.tsx (hunk 2 above)
Comment (yevhenkravets) [reply stencil-workspace/src/components/modus-date-picker/modus-date-picker.tsx:176]: Sure, thank you!

**Comment by yevhenkravets** [comment]
After the discussion with Carlos and the Modus team, we decided to close the ticket and PR.
@cjwinsor The Issues you found will be fixed in another PR.

**Pull request closed without merging**